// @tensorflow/tfjs-layers/dist-es6/utils/test_utils.js
import { memory, test_util } from '@tensorflow/tfjs-core';
import { describeWithFlags } from '@tensorflow/tfjs-core/dist/jasmine_util';
import { disposeScalarCache } from '../backend/tfjs_backend';
import { ValueError } from '../errors';
export function expectTensorsClose(actual, expected, epsilon) {
    if (actual == null) {
        throw new ValueError('First argument to expectTensorsClose() is not defined.');
    }
    if (expected == null) {
        throw new ValueError('Second argument to expectTensorsClose() is not defined.');
    }
    test_util.expectArraysClose(actual, expected, epsilon);
}
export function expectTensorsValuesInRange(actual, low, high) {
    if (actual == null) {
        throw new ValueError('First argument to expectTensorsClose() is not defined.');
    }
    test_util.expectValuesInRange(actual.dataSync(), low, high);
}
export function describeMathCPUAndGPU(testName, tests) {
    describeWithFlags(testName, {}, function () {
        beforeEach(function () {
            disposeScalarCache();
        });
        tests();
    });
}
export function describeMathCPU(testName, tests) {
    describeWithFlags(testName, test_util.CPU_ENVS, function () {
        beforeEach(function () {
            disposeScalarCache();
        });
        tests();
    });
}
export function describeMathGPU(testName, tests) {
    describeWithFlags(testName, test_util.WEBGL_ENVS, function () {
        beforeEach(function () {
            disposeScalarCache();
        });
        tests();
    });
}
export function expectNoLeakedTensors(testFunc, numNewTensors) {
    testFunc();
    var numTensorsBefore = memory().numTensors;
    testFunc();
    var numTensorsAfter = memory().numTensors;
    var actualNewTensors = numTensorsAfter - numTensorsBefore;
    if (actualNewTensors !== numNewTensors) {
        throw new ValueError("Created an unexpected number of new " +
            ("Tensors.  Expected: " + numNewTensors + ", created : " + actualNewTensors + ". ") +
            "Please investigate the discrepency and/or use tidy.");
    }
}
